modules/imap/functions.php

```diff
@@ -1417,3 +1417,79 @@ function parse_sieve_config_host($host) {
     // $host = '$scheme://'.$host;
     return [$host, $port, $tls];
 }}
+
+if (!hm_exists('connect_to_imap_serve')) {
+    function connect_to_imap_server($address, $name, $port, $user, $pass, $tls, $imap_sieve_host, $enableSieve, $context, $errno = null, $errstr = null) {
+        if ($con = fsockopen($address, $port, $errno, $errstr, 5)) {
+              $imap_list = array(
+                  'name' => $name,
+                  'server' => $address,
+                  'hide' => false,
+                  'port' => $port,
+                  'user' => $user,
+                  'pass' => $pass,
+                  'tls' => $tls);
+                  
+              Hm_IMAP_List::add($imap_list);
+              $servers = Hm_IMAP_List::dump(false, true);
+              $ids = array_keys($servers);
+              $imap_server_id = array_pop($ids);
+              
+             return $imap_server_id;
+        }
+    }
+}
+
+
+if (!hm_exists('connect_to_jmap_serve')) {
+    function connect_to_jmap_server($jmap_address, $hide_from_c_page, $name, $user, $pass) {
+       $hidden = false;
+                       
+       if($hide_from_c_page) {
+           $hidden = true;
+       }
+       
+       $parsed = parse_url($jmap_address);
+       
+       if (array_key_exists('host', $parsed) && @get_headers($jmap_address)) {
```

Comment:
hm-jmap should be used here as well to abstract the connection to the jmap server